File: Cocoa/Plug-Ins/Template Editor/ElementList.m
#import "ElementList.h"

#import "ElementDBYT.h"
#import "ElementDWRD.h"
#import "ElementDLNG.h"
#import "ElementDLLG.h"
#import "ElementUBYT.h"
#import "ElementUWRD.h"
#import "ElementULNG.h"
#import "ElementULLG.h"
#import "ElementHBYT.h"
#import "ElementHWRD.h"
#import "ElementHLNG.h"
#import "ElementHLLG.h"
#import "ElementREAL.h"
#import "ElementDOUB.h"
#import "ElementFIXD.h"
#import "ElementFRAC.h"
#import "ElementAWRD.h"
#import "ElementFBYT.h"
#import "ElementPSTR.h"
#import "ElementCHAR.h"
#import "ElementTNAM.h"
#import "ElementBOOL.h"
#import "ElementBFLG.h"
#import "ElementWFLG.h"
#import "ElementLFLG.h"
#import "ElementBBIT.h"
#import "ElementWBIT.h"
#import "ElementLBIT.h"
#import "ElementRECT.h"
#import "ElementPNT.h"
#import "ElementHEXD.h"
#import "ElementDATE.h"
#import "ElementOCNT.h"
#import "ElementFCNT.h"
#import "ElementLSTB.h"
#import "ElementKEYB.h"
#import "ElementCASE.h"
#import "ElementKRID.h"
#import "ElementRSID.h"
#import "ElementCOLR.h"

@implementation ElementList

+ (instancetype)listFromStream:(NSInputStream *)stream
{
    return [[self alloc] initFromStream:stream];
}

- (instancetype)initFromStream:(NSInputStream *)stream
{
    self = [super init];
    if (!self) return nil;
    self.parsed = NO;
    self.elements = [NSMutableArray new];
    [stream open];
    Element *element;
    while ([stream hasBytesAvailable]) {
        element = [self readElementFromStream:stream];
        if (!element) break;
        [self.elements addObject:element];
    }
    [stream close];
    return self;
}

- (id)copyWithZone:(NSZone *)zone
{
    ElementList *list = [[ElementList allocWithZone:zone] init];
    list.parsed = NO;
    list.controller = self.controller;
    list.elements = [[NSMutableArray allocWithZone:zone] initWithArray:self.elements copyItems:YES];
    [list parseElements];
    return list;
}

- (void)parseElements
{
    if (self.parsed) return;
    _currentIndex = 0;
    self.visibleElements = [NSMutableArray new];
    Element *element;
    while (_currentIndex < self.elements.count) {
        element = self.elements[_currentIndex];
        element.parentList = self;
        if (element.visible)
            [self.visibleElements addObject:element];
        [element readSubElements];
        _currentIndex++;
    }
    self.parsed = YES;
}

#pragma mark -

- (NSUInteger)count
{
    return self.visibleElements.count;
}

- (Element *)elementAtIndex:(NSUInteger)index
{
    return self.visibleElements[index];
}

// Insert a new element at the current position
- (void)insertElement:(Element *)element
{
    element.parentList = self;
    if (!self.parsed) {
        // Insert after current element during parsing (e.g. fixed count list, keyed section)
        [self.elements insertObject:element atIndex:++_currentIndex];
        [self.visibleElements addObject:element];
    } else {
        // Insert after current element during reading data (e.g. other lists)
        [self.elements insertObject:element atIndex:_currentIndex++];
        NSUInteger visIndex = [self.visibleElements indexOfObject:self.elements[_currentIndex]];
        [self.visibleElements insertObject:element atIndex:visIndex];
    }
}

// Insert a new element before/after a given element
- (void)insertElement:(Element *)element before:(Element *)before
{
    [self.elements insertObject:element atIndex:[self.elements indexOfObject:before]];
    element.parentList = self;
    [self.visibleElements insertObject:element atIndex:[self.visibleElements indexOfObject:before]];
}

- (void)insertElement:(Element *)element after:(Element *)after
{
    [self.elements insertObject:element atIndex:[self.elements indexOfObject:after]+1];
    element.parentList = self;
    [self.visibleElements insertObject:element atIndex:[self.visibleElements indexOfObject:after]+1];
}

- (void)removeElement:(Element *)element
{
    [self.elements removeObject:element];
    [self.visibleElements removeObject:element];
}

#pragma mark -

// The following methods may be used by elements while reading their sub elements

// Peek at an element in the list without removing it
- (Element *)peek:(NSUInteger)n
{
    n += _currentIndex;
    if (n >= self.elements.count)
        return nil;
    return self.elements[n];
}

// Pop the next element out of the list
- (Element *)pop
{
    if (_currentIndex+1 >= self.elements.count)
        return nil;
    Element *element = self.elements[_currentIndex+1];
    [self.elements removeObjectAtIndex:_currentIndex+1];
    return element;
}

// Search for an element of a given type following the current one
- (Element *)nextOfType:(NSString *)type
{
    Element *element;
    for (NSUInteger i = _currentIndex+1; i < self.elements.count; i++) {
        element = self.elements[i];
        if (!element) {
            NSLog(@"Element of type '%@' not found in template.", type);
            break;
        }
        if ([element.type isEqualToString:type])
            return element;
    }
    return nil;
}

// Create a new ElementList by extracting all elements following the current one up until a given type
- (ElementList *)subListFrom:(Element *)startElement
{
    ElementList *list = [ElementList new];
    list.controller = self.controller;
    list.elements = [NSMutableArray new];
    NSUInteger nesting = 0;
    Element *element;
    while (true) {
        element = [self pop];
        if (!element) {
            NSLog(@"Closing '%@' element not found for opening '%@'.", startElement.endType, startElement.type);
            break;
        }
        if ([element.endType isEqualToString:startElement.endType]) {
            nesting++;
        } else if ([element.type isEqualToString:startElement.endType]) {
            if (!nesting) break;
            nesting--;
        }
        [list.elements addObject:element];
    };
    return list;
}

#pragma mark -

- (void)readDataFrom:(ResourceStream *)stream
{
    _currentIndex = 0;
    // Don't use fast enumeration here as the list may be modified while reading
    while (_currentIndex < self.elements.count) {
        if (![stream bytesToGo]) return;
        [self.elements[_currentIndex] readDataFrom:stream];
        _currentIndex++;
    }
}

- (void)sizeOnDisk:(UInt32 *)size
{
    for (Element *element in self.elements) {
        [element sizeOnDisk:size];
    }
}

- (void)writeDataTo:(ResourceStream *)stream
{
    for (Element *element in self.elements) {
        [element writeDataTo:stream];
    }
}

#pragma mark -

- (Element *)readElementFromStream:(NSInputStream *)stream
{
    uint8_t sLength = 0;
    [stream read:&sLength maxLength:1];
    uint8_t *lBuffer = malloc(sLength);
    [stream read:lBuffer maxLength:sLength];
    NSString *label = [[NSString alloc] initWithBytesNoCopy:lBuffer length:sLength encoding:NSMacOSRomanStringEncoding freeWhenDone:YES];
    uint8_t *tBuffer = malloc(4);
    NSInteger bytesRead = [stream read:tBuffer maxLength:4];
    NSString *type = [[NSString alloc] initWithBytesNoCopy:tBuffer length:4 encoding:NSMacOSRomanStringEncoding freeWhenDone:YES];
    if (bytesRead != 4) {
        NSLog(@"Corrupt TMPL resource: not enough data.");
        return nil;
    }
    
    // create element class
    Class class = [self fieldRegistry][type];
    // check for Xnnn type - currently using resorcerer's nmm restriction to reserve all alpha types for potential future use
    if (!class && [type rangeOfString:@"[A-Z][0-9][0-9A-F]{2}" options:NSRegularExpressionSearch].location != NSNotFound) {
        class = [self fieldRegistry][[type substringToIndex:1]];
    }
    // check for XXnn type
    if (!class && [type rangeOfString:@"[A-Z]{2}[0-9]{2}" options:NSRegularExpressionSearch].location != NSNotFound) {
        class = [self fieldRegistry][[type substringToIndex:2]];
    }
    if (!class) {
        NSLog(@"Unrecognized template element type '%@'.", type);
        return nil;
    }
    return (Element *)[class elementForType:type withLabel:label];
}

- (NSMutableDictionary *)fieldRegistry
{
    static NSMutableDictionary *registry = nil;
    if (!registry) {
        registry = [[NSMutableDictionary alloc] init];
        
        // integers
        registry[@"DBYT"] = [ElementDBYT class];    // signed ints
        registry[@"DWRD"] = [ElementDWRD class];
        registry[@"DLNG"] = [ElementDLNG class];
        registry[@"DLLG"] = [ElementDLLG class];
        registry[@"UBYT"] = [ElementUBYT class];    // unsigned ints
        registry[@"UWRD"] = [ElementUWRD class];
        registry[@"ULNG"] = [ElementULNG class];
        registry[@"ULLG"] = [ElementULLG class];
        registry[@"HBYT"] = [ElementHBYT class];    // hex byte/word/long
        registry[@"HWRD"] = [ElementHWRD class];
        registry[@"HLNG"] = [ElementHLNG class];
        registry[@"HLLG"] = [ElementHLLG class];
        
        // multiple fields
        registry[@"RECT"] = [ElementRECT class];    // QuickDraw rect
        registry[@"PNT "] = [ElementPNT  class];    // QuickDraw point
        
        // align & fill
        registry[@"AWRD"] = [ElementAWRD class];    // alignment ints
        registry[@"ALNG"] = [ElementAWRD class];
        registry[@"AL08"] = [ElementAWRD class];
        registry[@"AL16"] = [ElementAWRD class];
        registry[@"FBYT"] = [ElementFBYT class];    // filler ints
        registry[@"FWRD"] = [ElementFBYT class];
        registry[@"FLNG"] = [ElementFBYT class];
        registry[@"FLLG"] = [ElementFBYT class];
        registry[@"F"]    = [ElementFBYT class];    // Fnnn
        
        // fractions
        registry[@"REAL"] = [ElementREAL class];    // single precision float
        registry[@"DOUB"] = [ElementDOUB class];    // double precision float
        registry[@"FIXD"] = [ElementFIXD class];    // 16.16 fixed fraction
        registry[@"FRAC"] = [ElementFRAC class];    // 2.30 fixed fraction
        
        // strings
        registry[@"PSTR"] = [ElementPSTR class];
        registry[@"BSTR"] = [ElementPSTR class];
        registry[@"WSTR"] = [ElementPSTR class];
        registry[@"LSTR"] = [ElementPSTR class];
        registry[@"OSTR"] = [ElementPSTR class];
        registry[@"ESTR"] = [ElementPSTR class];
        registry[@"CSTR"] = [ElementPSTR class];
        registry[@"OCST"] = [ElementPSTR class];
        registry[@"ECST"] = [ElementPSTR class];
        registry[@"P"]    = [ElementPSTR class];    // Pnnn
        registry[@"C"]    = [ElementPSTR class];    // Cnnn
        registry[@"CHAR"] = [ElementCHAR class];
        registry[@"TNAM"] = [ElementTNAM class];
        
        // bits
        registry[@"BOOL"] = [ElementBOOL class];    // true = 256; false = 0
        registry[@"BFLG"] = [ElementBFLG class];    // binary flag the size of a byte/word/long
        registry[@"WFLG"] = [ElementWFLG class];
        registry[@"LFLG"] = [ElementLFLG class];
        registry[@"BBIT"] = [ElementBBIT class];    // bit within a byte
        registry[@"BB"]   = [ElementBBIT class];    // BBnn bit field
        registry[@"BF"]   = [ElementBBIT class];    // BFnn fill bits
        registry[@"WBIT"] = [ElementWBIT class];
        registry[@"WB"]   = [ElementWBIT class];    // WBnn
        registry[@"WF"]   = [ElementWBIT class];    // WFnn
        registry[@"LBIT"] = [ElementLBIT class];
        registry[@"LB"]   = [ElementLBIT class];    // LBnn
        registry[@"LF"]   = [ElementLBIT class];    // LFnn
        
        // hex dumps
        registry[@"BHEX"] = [ElementHEXD class];
        registry[@"WHEX"] = [ElementHEXD class];
        registry[@"LHEX"] = [ElementHEXD class];
        registry[@"BSHX"] = [ElementHEXD class];
        registry[@"WSHX"] = [ElementHEXD class];
        registry[@"LSHX"] = [ElementHEXD class];
        registry[@"HEXD"] = [ElementHEXD class];
        registry[@"H"]    = [ElementHEXD class];    // Hnnn
        
        // list counters
        registry[@"OCNT"] = [ElementOCNT class];
        registry[@"ZCNT"] = [ElementOCNT class];
        registry[@"BCNT"] = [ElementOCNT class];
        registry[@"BZCT"] = [ElementOCNT class];
        registry[@"WCNT"] = [ElementOCNT class];
        registry[@"WZCT"] = [ElementOCNT class];
        registry[@"LCNT"] = [ElementOCNT class];
        registry[@"LZCT"] = [ElementOCNT class];
        registry[@"FCNT"] = [ElementFCNT class];    // fixed count with count in label (why not Lnnn?)
        // list begin/end
        registry[@"LSTB"] = [ElementLSTB class];
        registry[@"LSTZ"] = [ElementLSTB class];
        registry[@"LSTC"] = [ElementLSTB class];
        registry[@"LSTE"] = [Element     class];
        
        // option lists
        registry[@"CASE"] = [ElementCASE class];    // single option for preceding element
        registry[@"RSID"] = [ElementRSID class];    // resouce id (signed word) - type and offset in label
        
        // key selection
        registry[@"KBYT"] = [ElementDBYT class];    // signed keys
        registry[@"KWRD"] = [ElementDWRD class];
        registry[@"KLNG"] = [ElementDLNG class];
        registry[@"KLLG"] = [ElementDLLG class];
        registry[@"KUBT"] = [ElementUBYT class];    // unsigned keys
        registry[@"KUWD"] = [ElementUWRD class];
        registry[@"KULG"] = [ElementULNG class];
        registry[@"KULL"] = [ElementULLG class];
        registry[@"KHBT"] = [ElementHBYT class];    // hex keys
        registry[@"KHWD"] = [ElementHWRD class];
        registry[@"KHLG"] = [ElementHLNG class];
        registry[@"KHLL"] = [ElementHLLG class];
        registry[@"KCHR"] = [ElementCHAR class];    // keyed MacRoman values
        registry[@"KTYP"] = [ElementTNAM class];
        registry[@"KRID"] = [ElementKRID class];    // key on ID of the resource
        // keyed section begin/end
        registry[@"KEYB"] = [ElementKEYB class];
        registry[@"KEYE"] = [Element     class];
        
        // dates
        registry[@"DATE"] = [ElementDATE class];    // 4-byte date (seconds since 1 Jan 1904)
        registry[@"MDAT"] = [ElementDATE class];
        
        // colours
        registry[@"COLR"] = [ElementCOLR class];    // 6-byte QuickDraw colour
        registry[@"WCOL"] = [ElementCOLR class];    // 2-byte (15-bit) colour (Rezilla) - These appear to be EV Nova-specific additions
        registry[@"LCOL"] = [ElementCOLR class];    // 4-byte (24-bit) colour (Rezilla)
        
        registry[@"DVDR"] = [Element     class];    // divider
        
        // and some faked ones just to increase compatibility (these are marked 'x' in the docs)
        registry[@"SFRC"] = [ElementUWRD class];    // 0.16 fixed fraction
        registry[@"FXYZ"] = [ElementUWRD class];    // 1.15 fixed fraction
        registry[@"FWID"] = [ElementUWRD class];    // 4.12 fixed fraction
        registry[@"TITL"] = [Element     class];    // resource title (e.g. utxt would have "Unicode Text"; must be first element of template, and not anywhere else)
        registry[@"CMNT"] = [Element     class];
        registry[@"LLDT"] = [ElementULLG class];    // 8-byte date (LongDateTime; seconds since 1 Jan 1904)
        registry[@"STYL"] = [ElementDBYT class];    // QuickDraw font style
        registry[@"SCPC"] = [ElementDWRD class];    // MacOS script code (ScriptCode)
        registry[@"LNGC"] = [ElementDWRD class];    // MacOS language code (LangCode)
        registry[@"RGNC"] = [ElementDWRD class];    // MacOS region code (RegionCode)
        registry[@"BORV"] = [ElementHBYT class];    // Multiple select OR-value byte/word/long (Rezilla)
        registry[@"WORV"] = [ElementHWRD class];
        registry[@"LORV"] = [ElementHLNG class];
    }
    return registry;
}

@end

File: Plug-Ins/Template Editor/ElementList.swift
import Cocoa
import RFSupport

class ElementList {
    private(set) weak var parentElement: Element?
    private(set) weak var controller: TemplateWindowController!
    private var elements: [Element] = []
    private var visibleElements: [Element] = []
    private var currentIndex = 0
    private var configured = false
    var count: Int {
        return visibleElements.count
    }
    
    init(controller: TemplateWindowController, parent: Element? = nil) {
        parentElement = parent
        self.controller = controller
    }
    
    func copy() throws -> ElementList {
        let list = ElementList(controller: controller, parent: parentElement)
        list.elements = elements.map({ $0.copy() as Element })
        try list.configure()
        return list
    }
    
    func configure() throws {
        guard !configured else {
            return
        }
        while currentIndex < elements.count {
            let element = elements[currentIndex]
            element.parentList = self
            if element.visible {
                visibleElements.append(element)
            }
            try element.configure()
            currentIndex += 1
        }
        configured = true
    }
    
    func readTemplate(data: Data) -> Bool {
        do {
            let reader = BinaryDataReader(data)
            while reader.position < reader.data.endIndex {
                if let element = try self.readElement(from: reader) {
                    elements.append(element)
                }
            }
            try self.configure()
            return true
        } catch let error {
            let element = ElementDVDR(type: "DVDR", label: "Template Error\n\(error.localizedDescription)")
            elements = [element]
            visibleElements = [element]
            return false
        }
    }
    
    func getResourceData() -> Data {
        let writer = BinaryDataWriter()
        self.writeData(to: writer)
        return writer.data
    }
    
    // MARK: -
    
    func element(at index: Int) -> Element {
        return visibleElements[index]
    }

    // Insert a new element before the current element during read/configure
    func insert(_ element: Element) {
        elements.insert(element, at: currentIndex)
        element.parentList = self
        // If the element list was previously empty then we haven't started configure yet
        if elements.count > 1 {
            currentIndex += 1
            if element.visible {
                let visIndex = visibleElements.firstIndex(of: elements[currentIndex]) ?? visibleElements.endIndex
                visibleElements.insert(element, at: visIndex)
            }
        }
    }
    
    // Insert a new element before/after a given element
    func insert(_ element: Element, before: Element) {
        elements.insert(element, at: elements.firstIndex(of: before)!)
        element.parentList = self
        if element.visible {
            visibleElements.insert(element, at: visibleElements.firstIndex(of: before)!)
        }
    }
    
    func insert(_ element: Element, after: Element) {
        elements.insert(element, at: elements.firstIndex(of: after)! + 1)
        element.parentList = self
        if element.visible {
            visibleElements.insert(element, at: visibleElements.firstIndex(of: element)! + 1)
        }
    }
    
    func remove(_ element: Element) {
        elements.removeAll(where: { $0 == element })
        visibleElements.removeAll(where: { $0 == element })
    }
    
    // MARK: -
    
    // The following methods may be used by elements while reading their sub elements
    
    // Peek at an element in the list without removing it
    func peek(_ n: Int) -> Element? {
        let i = currentIndex + n
        return i < elements.endIndex ? elements[i] : nil
    }
    
    // Pop the next element out of the list
    func pop(_ type: String? = nil) -> Element? {
        let i = currentIndex + 1
        if i >= elements.endIndex {
            return nil
        }
        let element = elements[i]
        if let type = type, element.type != type {
            return nil
        }
        elements.remove(at: i)
        return element
    }
    
    // Search for an element of a given type following the current one
    func next(ofType type: String) -> Element? {
        return elements[(currentIndex+1)...].first(where: { $0.type == type })
    }
    
    // Search for an element of a given type preceding the current one, travsersing up the hierarchy if necessary
    func previous(ofType type: String) -> Element? {
        if let element = elements[0..<currentIndex].last(where: { $0.type == type }) {
            return element
        }
        if let parent = parentElement {
            return parent.parentList.previous(ofType: type)
        }
        return nil
    }
    
    // Search for the next visible element with the given display label
    func next(withLabel label: String) -> Element? {
        for el in elements[(currentIndex+1)...] where el.visible {
            if el.displayLabel == label {
                return el
            }
        }
        return nil
    }
    
    // Create a new ElementList by extracting all elements following the current one up until a given type
    func subList(for startElement: Element) throws -> ElementList {
        let list = ElementList(controller: controller, parent: startElement)
        var nesting = 0
        while true {
            guard let element = self.pop() else {
                throw TemplateError.unclosedElement(startElement)
            }
            if element.endType == startElement.endType {
                nesting += 1
            } else if element.type == startElement.endType {
                if nesting == 0 {
                    break
                }
                nesting -= 1
            }
            list.elements.append(element)
        }
        return list
    }
    
    // MARK: -
    
    func readData(from reader: BinaryDataReader) throws {
        let bigEndian = reader.bigEndian
        currentIndex = 0;
        // Don't use for loop here as the list may be modified while reading
        while currentIndex < elements.count && reader.position < reader.data.endIndex {
            try elements[currentIndex].readData(from: reader)
            currentIndex += 1
        }
        // Always restore original endianness at the end of any sublist
        reader.bigEndian = bigEndian
    }
    
    func writeData(to writer: BinaryDataWriter) {
        let bigEndian = writer.bigEndian
        for element in elements {
            element.writeData(to: writer)
        }
        writer.bigEndian = bigEndian
    }
    
    // MARK: -
    
    func readElement(from reader: BinaryDataReader) throws -> Element? {
        guard let label = try? reader.readPString(),
              let typeCode: FourCharCode = try? reader.read()
        else {
            throw TemplateError.corrupt
        }
        let type = typeCode.stringValue
        
        var elType = Self.fieldRegistry[type]
        
        // check for Xnnn type - currently using resorcerer's nmm restriction to reserve all-alpha types (e.g. FACE) for potential future use
        if elType == nil && type.range(of: "[A-Z](?!000)[0-9][0-9A-F]{2}", options: .regularExpression) != nil {
            if type.first == "R" {
                // Rnnn psuedo-element repeats the following element n times
                let count = Int(type.suffix(3), radix: 16)!
                let offset = Int(label.split(separator: "=").last!) ?? 1
                guard var el = try self.readElement(from: reader) else {
                    throw TemplateError.corrupt
                }
                let l = el.label
                for i in 0..<count {
                    // Replace * symbol with index
                    let label = l.replacingOccurrences(of: "*", with: String(i+offset))
                    el = Swift.type(of: el).init(type: el.type, label: label, tooltip: el.tooltip)!
                    elements.append(el)
                }
                return nil
            }
            elType = Self.fieldRegistry[String(type.prefix(1))]
        }
        
        // check for XXnn type
        if elType == nil && type.range(of: "[A-Z]{2}(?!00)[0-9]{2}", options: .regularExpression) != nil {
            elType = Self.fieldRegistry[String(type.prefix(2))]
        }
        if let elType = elType, let el = elType.init(type: type, label: label) {
            return el
        } else {
            throw TemplateError.unknownElement(type)
        }
    }
    
    static let fieldRegistry: [String: Element.Type] = [
        // integers
        "DBYT": ElementDBYT<Int8>.self,     // signed ints
        "DWRD": ElementDBYT<Int16>.self,
        "DLNG": ElementDBYT<Int32>.self,
//        "DLLG": ElementDBYT<Int64>.self,    // (ResForge)
        "UBYT": ElementDBYT<UInt8>.self,    // unsigned ints
        "UWRD": ElementDBYT<UInt16>.self,
        "ULNG": ElementDBYT<UInt32>.self,
//        "ULLG": ElementDBYT<UInt64>.self,   // (ResForge)
        "HBYT": ElementHBYT<UInt8>.self,    // hex byte/word/long
        "HWRD": ElementHBYT<UInt16>.self,
        "HLNG": ElementHBYT<UInt32>.self,
//        "HLLG": ElementHBYT<UInt64>.self,   // (ResForge)

        // multiple fields
        "RECT": ElementRECT.self,           // QuickDraw rect
        "PNT ": ElementPNT.self,            // QuickDraw point

        // align & fill
        "AWRD": ElementAWRD.self,           // alignment ints
        "ALNG": ElementAWRD.self,
        "AL08": ElementAWRD.self,
        "AL16": ElementAWRD.self,
        "FBYT": ElementFBYT.self,           // filler ints
        "FWRD": ElementFBYT.self,
        "FLNG": ElementFBYT.self,
        "FLLG": ElementFBYT.self,
        "F"   : ElementFBYT.self,           // Fnnn

        // fractions
        "REAL": ElementREAL.self,           // single precision float
        "DOUB": ElementDOUB.self,           // double precision float

        // strings
        "PSTR": ElementPSTR<UInt8>.self,
        "BSTR": ElementPSTR<UInt8>.self,
        "WSTR": ElementPSTR<UInt16>.self,
        "LSTR": ElementPSTR<UInt32>.self,
        "OSTR": ElementPSTR<UInt8>.self,
        "ESTR": ElementPSTR<UInt8>.self,
        "CSTR": ElementPSTR<UInt32>.self,
        "OCST": ElementPSTR<UInt32>.self,
        "ECST": ElementPSTR<UInt32>.self,
        "TXTS": ElementPSTR<UInt32>.self,
        "P"   : ElementPSTR<UInt8>.self,    // Pnnn
        "C"   : ElementPSTR<UInt32>.self,   // Cnnn
        "T"   : ElementPSTR<UInt32>.self,   // Tnnn
        "CHAR": ElementCHAR.self,
        "TNAM": ElementTNAM.self,

        // bits
        "BOOL": ElementBOOL.self,           // true = 256; false = 0
        "BFLG": ElementBFLG<UInt8>.self,    // binary flag the size of a byte/word/long
        "WFLG": ElementBFLG<UInt16>.self,
        "LFLG": ElementBFLG<UInt32>.self,
        "BBIT": ElementBBIT<UInt8>.self,    // bit within a byte
        "BB"  : ElementBBIT<UInt8>.self,    // BBnn bit field
        "BF"  : ElementBBIT<UInt8>.self,    // BFnn fill bits (ResForge)
        "WBIT": ElementBBIT<UInt16>.self,
        "WB"  : ElementBBIT<UInt16>.self,   // WBnn
        "WF"  : ElementBBIT<UInt16>.self,   // WFnn (ResForge)
        "LBIT": ElementBBIT<UInt32>.self,
        "LB"  : ElementBBIT<UInt32>.self,   // LBnn
        "LF"  : ElementBBIT<UInt32>.self,   // LFnn (ResForge)
        "BORV": ElementBORV<UInt8>.self,    // OR-value (Rezilla)
        "WORV": ElementBORV<UInt16>.self,
        "LORV": ElementBORV<UInt32>.self,

        // hex dumps
        "HEXD": ElementHEXD.self,
        "HEXS": ElementHEXD.self,
        "H"   : ElementHEXD.self,           // Hnnn
        "BHEX": ElementBHEX<UInt8>.self,
        "WHEX": ElementBHEX<UInt16>.self,
        "LHEX": ElementBHEX<UInt32>.self,
        "BSHX": ElementBHEX<UInt8>.self,
        "WSHX": ElementBHEX<UInt16>.self,
        "LSHX": ElementBHEX<UInt32>.self,

        // list counters
        "BCNT": ElementOCNT<UInt8>.self,
        "OCNT": ElementOCNT<UInt16>.self,
        "LCNT": ElementOCNT<UInt32>.self,
        "ZCNT": ElementOCNT<Int16>.self,
        "LZCT": ElementOCNT<Int32>.self,
        "FCNT": ElementFCNT.self,           // fixed count with count in label (why didn't they choose Lnnn?)
        "R"   : Element.self,               // single-element repeat (ResForge) (never initialised but included here for reference)
        // list begin/end
        "LSTB": ElementLSTB.self,
        "LSTS": ElementLSTB.self,
        "LSTZ": ElementLSTB.self,
        "LSTC": ElementLSTB.self,
        "LSTE": Element.self,

        // option lists
        "CASE": ElementCASE.self,           // single option for preceding element
        "CASR": ElementCASR.self,           // option range for preceding element (ResForge)
        "RSID": ElementRSID.self,           // resouce id (signed word) - type and offset in label

        // key selectors
        "KBYT": ElementKBYT<Int8>.self,     // signed keys
        "KWRD": ElementKBYT<Int16>.self,
        "KLNG": ElementKBYT<Int32>.self,
//        "KLLG": ElementKBYT<Int64>.self,    // (ResForge)
        "KUBT": ElementKBYT<UInt8>.self,    // unsigned keys
        "KUWD": ElementKBYT<UInt16>.self,
        "KULG": ElementKBYT<UInt32>.self,
//        "KULL": ElementKBYT<UInt64>.self,   // (ResForge)
        "KHBT": ElementKHBT<UInt8>.self,    // hex keys
        "KHWD": ElementKHBT<UInt16>.self,
        "KHLG": ElementKHBT<UInt32>.self,
//        "KHLL": ElementKHBT<UInt64>.self,   // (ResForge)
        "KCHR": ElementKCHR.self,           // string keys
        "KTYP": ElementKTYP.self,
        "KRID": ElementKRID.self,           // key on ID of the resource
        // keyed section begin/end
        "KEYB": ElementKEYB.self,
        "KEYE": Element.self,

        // dates
        "DATE": ElementDATE.self,           // 4-byte date (seconds since 1 Jan 1904)
        "MDAT": ElementDATE.self,

        // colours
        "COLR": ElementCOLR.self,           // 6-byte QuickDraw colour
        "WCOL": ElementWCOL<UInt16>.self,   // 2-byte (15-bit) colour (Rezilla)
        "LCOL": ElementWCOL<UInt32>.self,   // 4-byte (24-bit) colour (Rezilla)
        
        // offsets
        "BSKP": ElementBSKP<UInt8>.self,
        "WSKP": ElementBSKP<UInt16>.self,
        "SKIP": ElementBSKP<UInt16>.self,
        "LSKP": ElementBSKP<UInt32>.self,
        "BSIZ": ElementBSKP<UInt8>.self,
        "WSIZ": ElementBSKP<UInt16>.self,
        "LSIZ": ElementBSKP<UInt32>.self,
        "SKPE": Element.self,
        
        // byte order
        "BIGE": ElementBNDN.self,           // Big-endian
        "LTLE": ElementBNDN.self,           // Little-endian
        "BNDN": ElementBNDN.self,           // Big-endian (hidden)
        "LNDN": ElementBNDN.self,           // Little-endian (hidden)

        // cosmetic
        "DVDR": ElementDVDR.self,           // divider
        "RREF": ElementRREF.self,           // static reference to another resource (ResForge)
        "PACK": ElementPACK.self,           // pack other elements together (ResForge)

        // and some faked ones just to increase compatibility
        "SFRC": ElementDBYT<UInt16>.self,   // 0.16 fixed fraction
        "FXYZ": ElementDBYT<UInt16>.self,   // 1.15 fixed fraction
        "FWID": ElementDBYT<UInt16>.self,   // 4.12 fixed fraction
        "FRAC": ElementDBYT<UInt32>.self,   // 2.30 fixed fraction
        "FIXD": ElementDBYT<UInt32>.self,   // 16.16 fixed fraction
        "SCPC": ElementDBYT<Int16>.self,    // MacOS script code (ScriptCode)
        "LNGC": ElementDBYT<Int16>.self,    // MacOS language code (LangCode)
        "RGNC": ElementDBYT<Int16>.self,    // MacOS region code (RegionCode)
        "CODE": ElementHEXD.self            // 680x0 Disassembled Code Dump
    ]
}
